oscore-cf/src/main/java/org/eclipse/californium/oscore/EncryptCommon.java

```diff
@@ -0,0 +1,213 @@
+            case AES_CCM_64_64_128:
+            case AES_CCM_64_64_256:
+            case AES_CCM_16_128_128:
+            case AES_CCM_16_128_256:
+            case AES_CCM_64_128_128:
+            case AES_CCM_64_128_256:
+                AES_CCM_Encrypt(alg, rgbKey);
+                break;
+
+            default:
+                throw new CoseException("Unsupported Algorithm Specified");
+        }
+    }
+    
+    private int getAES_CCM_IVSize(AlgorithmID alg) throws CoseException {
+        switch (alg) {
+            case AES_CCM_16_64_128:
+            case AES_CCM_16_64_256:
+            case AES_CCM_16_128_128:
+            case AES_CCM_16_128_256:
+                return AES_CCM_16_IV_LENGTH;
+            case AES_CCM_64_64_128:
+            case AES_CCM_64_64_256:
+            case AES_CCM_64_128_256:
+            case AES_CCM_64_128_128:
+                return AES_CCM_64_IV_LENGTH;
+		default:
+			 throw new CoseException("Unsupported Algorithm Specified");
+        }
+    }
+    
+    private void AES_CCM_Decrypt(AlgorithmID alg, byte[] rgbKey) throws CoseException, IllegalStateException {
+        // validate key
+        if (rgbKey.length != alg.getKeySize()/8) {
+            throw new CoseException("Key Size is incorrect");
+        }
+
+        // obtain and validate IV
+        final int ivLen = getAES_CCM_IVSize(alg);
+        CBORObject  iv = findAttribute(HeaderKeys.IV);
+        if (iv == null) {
+            throw new CoseException("Missing IV during decryption");
+        }
+        if (iv.getType() != CBORType.ByteString) {
+            throw new CoseException("IV is incorrectly formed");
+        }
+        if (iv.GetByteString().length != ivLen) {
+            throw new CoseException("IV size is incorrect");
+        }
+        
+        try {
+            rgbContent = CCMBlockCipher.decrypt(rgbKey, iv.GetByteString(), getExternal(), getEncryptedContent(), 0);
+        } catch (NoSuchAlgorithmException ex) {
+            throw new CoseException("Algorithm not supported", ex);
+        }
+        catch (InvalidKeyException ex) {
+            if (ex.getMessage() == "Illegal key size") {
+                throw new CoseException("Unsupported key size", ex);
+            }
+            throw new CoseException("Decryption failure", ex);
+        } catch (Exception ex) {
+        	ex.printStackTrace();
+            throw new CoseException("Decryption failure", ex);
+        }
+    }
+    
+ 
+    private void AES_CCM_Encrypt(AlgorithmID alg, byte[] rgbKey) throws CoseException, IllegalStateException {
+        // validate key
+        if (rgbKey.length != alg.getKeySize()/8) {
+            throw new CoseException("Key Size is incorrect");
+        }
+        
+        // obtain and validate iv
+        CBORObject  iv = findAttribute(HeaderKeys.IV);
+        int         ivLen = getAES_CCM_IVSize(alg);
+        if (iv == null) {
+            byte[] tmp = new byte[ivLen];
+            random.nextBytes(tmp);
```

Comment:
Is `AES_CCM_Encrypt` called more then once per Message (in order to reuse the `random`)?

Reply:
No. It is not. Do you mean the random should just be localized for the function and re-initiated for each time it is called? I can see your point, the change feels sound. There is no need to keep the variable alive.

Reply:
If it's called only once, it always re-initiated, either during the class constructor or during the function call. Therefore I would move it from the field inside to the function.

Reply:
Sounds reasonable.